include/bluetooth/services/dfu_smp_c.h

```diff
@@ -0,0 +1,192 @@
+ *
+ *  Current response state.
+ */
+struct bt_gatt_dfu_smp_rsp_state {
+	/**< @brief Total size of the response
+	 *
+	 *  Total expected size of the response that would be updated
+	 *  from the first part of the response, where SMP header is located.
+	 */
+	size_t total_size;
+	/**< @brief Current data chunk offset  */
+	size_t offset;
+	/**< @brief The size of the data chunk */
+	size_t chunk_size;
+	/**< @brief Pointer to the data        */
+	const u8_t *data;
+};
+
+/** @brief Handle part of the response
+ *
+ *  @param dfu_smp_c DFU SMP Client instance
+ */
+typedef void
+(*bt_gatt_dfu_smp_rsp_part_cb)(struct bt_gatt_dfu_smp_c *dfu_smp_c);
+
+/** @brief Global function called on error
+ *
+ *  @param dfu_smp_c DFU SMP Client instance
+ *  @param err       Negative internal error code or positive GATT error code.
+ */
+typedef void
+(*bt_gatt_dfu_smp_error_cb)(struct bt_gatt_dfu_smp_c *dfu_smp_c, int err);
+
+/**
+ * @brief DFU SMP client parameters for the initialization function.
+ */
+struct bt_gatt_dfu_smp_c_init_params {
+	/** @brief Callback pointer for error handler
+	 *
+	 *  Function called on error
+	 */
+	bt_gatt_dfu_smp_error_cb error_cb;
+};
+
+/**
+ * DFU SMP client structure
+ */
+struct bt_gatt_dfu_smp_c {
+	/** Connection object */
+	struct bt_conn *conn;
+	/** Handlers */
+	struct {
+		u16_t smp;     /**< SMP characteristic value handler */
+		u16_t smp_ccc; /**< SMP characteristic CCC handler */
+	} handles;
+	/** @brief Current response state.
+	 */
+	struct bt_gatt_dfu_smp_rsp_state rsp_state;
+	/** Callbacks */
+	struct {
+		/** @brief Callback pointer for global error
+		 *
+		 *  Function called when some global error appears
+		 */
+		bt_gatt_dfu_smp_error_cb error_cb;
+		/** @brief Callback pointer for response
+		 *
+		 * The pointer to callback function to receive response.
+		 * It is set to NULL when response is finished.
+		 */
+		bt_gatt_dfu_smp_rsp_part_cb rsp_part;
+	} cbs;
+	/** @brief Response notification parameters
+	 *
+	 *  Parameters set for the SMP response notification.
+	 *
+	 *  The status of the notification is reflected by @c notify field.
+	 *  Notification has to be enabled before command is sent.
+	 *  This is handled automatically by the library.
+	 */
+	struct bt_gatt_subscribe_params notification_params;
+};
+
+/** @brief Initialize the DFU SMP Client module
+ *
+ *  This function initializes the DFU SMP Client module.
+ *
+ *  @param[out] dfu_smp_c DFU SMP Client instance.
+ *  @param[in]  params Initialization parameter.
+ *
+ *  @return 0 If the operation was successful.
+ *          Otherwise a negative error code is returned.
+ */
+int bt_gatt_dfu_smp_c_init(struct bt_gatt_dfu_smp_c *dfu_smp_c,
+			   const struct bt_gatt_dfu_smp_c_init_params *params);
+
+/** @brief Assign handles to the DFU SMP Client instance.
+ *
+ *  @param[in,out] dm Discovery object.
+ *  @param[in,out] dfu_smp_c DFU SMP Client instance.
+ *
+ *  @retval 0 If the operation was successful.
```

Comment:
This one here is actually correct (except that the "Otherwise" case should be included in one of the retval).
retval should be used whenever there is a specific value, return when you describe what the value should be. For one function, retval and return should not be mixed.

Reply:
Here's the style guide for that part with the text that we usually use:
https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/doc_styleguide.html#functions

Reply:
I can see in the documentation mentioned above:

```
 If there is only one retval, add what happens otherwise. Example: “Otherwise, an error code is returned”
```

What if there is more than one retval?
The comma here looks odd.
Maybe it should be something like:

```
@retval Other Negative error code
```

It would go into table and should look nice.

Reply:
Your last suggestion would mean that the value is "Other", and it occurs in the case of "Negative error code".
Otherwise is a weird case, I admit, but it just needs to be combined with one of the other cases. I added it to the "standard" one now.

Reply:
This is something I used to use in the documentation in such cases:
```
@retval <0 Negative error code
```